File: GFG/Two numbers with odd occurrences/Two numbers with odd occurrences.cpp
class Solution{
    public:
    vector<long long int> twoOddNum(long long int Arr[], long long int N)  
    {
        // code here
        vector< long long int> v;

long long int res=0,x=0,count=0,b=0;

 

for(long long int i=0;i<N;i++)

{

  res=res^Arr[i];

}

while(!(res&(1<<count))){

 ++count;

}

b=b|(1<<count);

 

for(long long int i=0;i<N;i++)

{  

 if(Arr[i]&b)

    x=x^res^Arr[i];

}

 

v.push_back(max(x,res^x));

v.push_back(min(x,res^x));

 

return v;
    }
};

File: GFG/Two numbers with odd occurrences/Two numbers with odd occurrences.java
class Solution
{
    public int[] twoOddNum(int arr[], int N)
    {
        // code here
        int xor=0;
        for(int i:arr){
            xor^=i;
        }
        int setBit=0;
        for(int i=0;i<32;i++){
            int curr= 1<<i;
            if((xor&curr)!=0){
                setBit=curr;
                break;
            }
        }
        int first=0;
        int second=0;
        for(int i=0;i<N;i++){
            if((arr[i]&setBit)!=0){
                first^=arr[i];
            }else{
                second^=arr[i];
            }
        }
        int[]ans=new int[2];
        if(first<second){
            ans[0]=second;
            ans[1]=first;
        }else{
            ans[0]=first;
            ans[1]=second;
        }
        return ans;
    }
}
